mlir/compiler-course/rezantseva_mlir/rezantseva_mlir.cpp

```diff
@@ -0,0 +1,112 @@
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/SPIRV/IR/SPIRVOps.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/Visitors.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Tools/Plugins/PassPlugin.h"
+#include <algorithm>
+#include <vector>
+
+using namespace mlir;
+
+namespace {
+
+class MaxDepthPass
+    : public PassWrapper<MaxDepthPass, OperationPass<func::FuncOp>> {
+private:
+  bool isLoopOp(Operation *op) {
+    return isa<affine::AffineForOp, scf::ForOp, scf::WhileOp, scf::ForallOp,
+               spirv::LoopOp>(op);
+  }
+
+  bool isNestingOp(Operation *op) {
+    return isa<affine::AffineForOp, scf::ForOp, scf::WhileOp, scf::ForallOp,
+               spirv::LoopOp, affine::AffineIfOp, scf::IfOp,
+               spirv::SelectionOp>(op);
+  }
+
+  unsigned computeDepth(Operation *op, unsigned currentDepth = 0) {
+    unsigned maxDepth = currentDepth;
+
+    if (isNestingOp(op)) {
+      maxDepth = std::max(maxDepth, currentDepth + 1);
+    }
+
+    for (Region &region : op->getRegions()) {
+      for (Block &block : region) {
+        for (Operation &nestedOp : block) {
+          maxDepth = std::max(
+              maxDepth, computeDepth(&nestedOp,
+                                     currentDepth + (isNestingOp(op) ? 1 : 0)));
+        }
+      }
+    }
+
+    return maxDepth;
+  }
+
+public:
+  StringRef getArgument() const final {
+    return "MaxDepthPass_RezantsevaAnastasia_FIIT1_MLIR";
+  }
+
+  StringRef getDescription() const final {
+    return "It's a pass, that counts the max depth of region nests in each "
+           "loop.";
+  }
+
+  void runOnOperation() override {
+    func::FuncOp func = getOperation();
+    std::vector<unsigned> loopDepths;
+
+    func.walk([&](Operation *op) {
+      if (isLoopOp(op)) {
+        Operation *parent = op->getParentOp();
+        bool isNestedInLoop = false;
+        while (parent && !isa<func::FuncOp>(parent)) {
+          if (isLoopOp(parent)) {
+            isNestedInLoop = true;
+            break;
+          }
+          parent = parent->getParentOp();
+        }
+        if (!isNestedInLoop) {
+          unsigned depth = computeDepth(op);
+          if (depth > 0) {
+            loopDepths.push_back(depth);
+          }
+        }
+      }
```

Comment:
`WalkResult::skip()` result will break recursive visiting nested operations of current `op` and goes to the next op at current or upper level.
```suggestion
      if (isLoopOp(op)) {
        unsigned depth = computeDepth(op);
        if (depth > 0) {
          loopDepths.push_back(depth);
        }
        return WalkResult::skip();
      }
```

Reply:
Thanks for the suggested edits. I changed the implementation to include WalkResult::skip() to skip nested operations when a loop is detected.

Reply:
You are welcome